llvm/lib/SYCLLowerIR/SYCLPropagateAspectsUsage.cpp

```diff
@@ -525,26 +650,149 @@ void propagateAspectsThroughCG(Function *F, CallGraphTy &CG,
   AspectsSetTy LocalAspects;
   for (Function *Callee : It->second) {
     if (Visited.insert(Callee).second)
-      propagateAspectsThroughCG(Callee, CG, AspectsMap, Visited);
+      propagateAspectsThroughCG(Callee, CG, AspectsMap,
+                                FunctionsCalledUnconditionally, Visited);
 
     const auto &CalleeAspects = AspectsMap[Callee];
     LocalAspects.insert(CalleeAspects.begin(), CalleeAspects.end());
   }
+  if (FunctionsCalledUnconditionally.contains(F))
+    AspectsMap[F].insert(LocalAspects.begin(), LocalAspects.end());
+}
+
+using PathsContainingConditionalCallersTy =
+    std::vector<std::vector<Function *>>;
+
+void identifyPathsContainingConditionalCallers(
+    Function *F, const CallGraphTy &CG, std::vector<Function *> &CurrentPath,
+    PathsContainingConditionalCallersTy &Paths) {
+  const auto It = CG.find(F);
+  if (It == CG.end()) {
+    CurrentPath.push_back(F);
+    for (auto Func : CurrentPath)
+      if (Func->hasFnAttribute("sycl-call-if-on-device-conditionally")) {
+        Paths.push_back(CurrentPath);
+        break;
+      }
+    CurrentPath.pop_back();
+    return;
+  }
 
-  AspectsMap[F].insert(LocalAspects.begin(), LocalAspects.end());
+  CurrentPath.push_back(F);
+
+  for (Function *Callee : It->second) {
+    identifyPathsContainingConditionalCallers(Callee, CG, CurrentPath, Paths);
+  }
+
+  CurrentPath.pop_back();
+}
+
+void propagateConditionExpressionsAndConditionalAspectsThroughCG(
```

Comment:
Some explanatory comments should be added to explain the difference between simply propagating conditional expressions vs propagating both conditional expressions and aspects. It is not clear why do we need both and that's the difference between them.